listings/applications/simple_vault/src/simple_vault.cairo

```diff
@@ -53,49 +64,145 @@ pub mod SimpleVault {
+        IERC20Dispatcher as IERC20Dispatcher_token
+    };
+    use starknet::testing::{set_contract_address, set_account_contract_address};
+    use starknet::{
+        ContractAddress, SyscallResultTrait, syscalls::deploy_syscall, get_caller_address,
+        contract_address_const
+    };
+
+    fn deploy() -> (ISimpleVaultDispatcher, ContractAddress, IERC20Dispatcher_token) {
+        let _token_address: ContractAddress = contract_address_const::<'token_address'>();
+        let caller = contract_address_const::<'caller'>();
+
+        let (token_contract_address, _) = deploy_syscall(
+            erc_20::TEST_CLASS_HASH.try_into().unwrap(),
+            caller.into(),
+            array![caller.into(), 'myToken', '8', '1000'.into(), 'MYT'].span(),
+            false
+        )
+            .unwrap_syscall();
+
+        let (contract_address, _) = deploy_syscall(
+            SimpleVault::TEST_CLASS_HASH.try_into().unwrap(),
+            0,
+            array![token_contract_address.into()].span(),
+            false
+        )
+            .unwrap_syscall();
+
+        (
+            ISimpleVaultDispatcher { contract_address },
+            contract_address,
+            IERC20Dispatcher_token { contract_address: token_contract_address }
+        )
+    }
+
+    #[test]
+    fn test_deposit() {
+        let caller = contract_address_const::<'caller'>();
+        let (dispatcher, vault_address, token_dispatcher) = deploy();
+
+        // Approve the vault to transfer tokens on behalf of the caller
+        let amount: felt252 = 10.into();
+        token_dispatcher.approve(vault_address.into(), amount);
+        set_contract_address(caller);
+
+        // Deposit tokens into the vault
+        let amount: u256 = 10.into();
+        dispatcher.deposit(amount);
+
+        // Check balances and total supply
+        let balance_of_caller = dispatcher.user_balance_of(caller);
+        let total_supply = dispatcher.contract_total_supply();
+
+        assert(balance_of_caller == amount, 'Deposit failed');
+        assert(total_supply == amount, 'total supply mismatch');
+
+    }
+
+    #[test]
+    fn test_deposit_withdraw() {
+        let caller = contract_address_const::<'caller'>();
+        let (dispatcher, vault_address, token_dispatcher) = deploy();
+
+        // Approve the vault to transfer tokens on behalf of the caller
+        let amount: felt252 = 10.into();
+        token_dispatcher.approve(vault_address.into(), amount);
+        set_contract_address(caller);
+        set_account_contract_address(vault_address);
+
+        // Deposit tokens into the vault
+        let amount: u256 = 10.into();
+        dispatcher.deposit(amount);
+        dispatcher.withdraw(amount);
+
+        // Check balances of user in the vault after withdraw
+        let balance_of_caller = dispatcher.user_balance_of(caller);
+
+        assert(balance_of_caller == 0.into(), 'withdraw failed');
+    }
```

Comment:
It would be nice to generate another test where the contract generate some yield after the user deposit then withdraw his tokens + yield by burning his shares.